staging/vhost-device-gpu/src/virtio_gpu.rs

```diff
@@ -0,0 +1,1055 @@
+        fence_id: u64,
+        desc_index: u16,
+        len: u32,
+    ) -> bool {
+        // In case the fence is signaled immediately after creation, don't add a return
+        // FenceDescriptor.
+        let mut fence_state = self.fence_state.lock().unwrap();
+        if fence_id > *fence_state.completed_fences.get(&ring).unwrap_or(&0) {
+            fence_state.descs.push(FenceDescriptor {
+                ring,
+                fence_id,
+                desc_index,
+                len,
+            });
+
+            false
+        } else {
+            true
+        }
+    }
+
+    fn resource_create_blob(
+        &mut self,
+        ctx_id: u32,
+        resource_id: u32,
+        resource_create_blob: ResourceCreateBlob,
+        vecs: Vec<(GuestAddress, usize)>,
+        mem: &GuestMemoryMmap,
+    ) -> VirtioGpuResult {
+        let mut rutabaga_iovecs = None;
+
+        if resource_create_blob.blob_flags & VIRTIO_GPU_BLOB_FLAG_CREATE_GUEST_HANDLE != 0 {
+            panic!("GUEST_HANDLE unimplemented");
+        } else if resource_create_blob.blob_mem != VIRTIO_GPU_BLOB_MEM_HOST3D {
+            rutabaga_iovecs =
+                Some(sglist_to_rutabaga_iovecs(&vecs[..], mem).map_err(|_| ErrUnspec)?);
+        }
+
+        self.rutabaga.resource_create_blob(
+            ctx_id,
+            resource_id,
+            resource_create_blob,
+            rutabaga_iovecs,
+            None,
+        )?;
+
+        let resource = VirtioGpuResource::new(resource_id, 0, 0, resource_create_blob.size);
+
+        // Rely on rutabaga to check for duplicate resource ids.
+        self.resources.insert(resource_id, resource);
+        Ok(self.result_from_query(resource_id))
+    }
+
+    fn resource_map_blob(
+        &mut self,
+        resource_id: u32,
+        shm_region: &VirtioShmRegion,
+        offset: u64,
+    ) -> VirtioGpuResult {
+        let resource = self
+            .resources
+            .get_mut(&resource_id)
+            .ok_or(ErrInvalidResourceId)?;
+
+        let map_info = self.rutabaga.map_info(resource_id).map_err(|_| ErrUnspec)?;
+
+        if let Ok(export) = self.rutabaga.export_blob(resource_id) {
+            if export.handle_type != RUTABAGA_MEM_HANDLE_TYPE_OPAQUE_FD {
+                let prot = match map_info & RUTABAGA_MAP_ACCESS_MASK {
+                    RUTABAGA_MAP_ACCESS_READ => libc::PROT_READ,
+                    RUTABAGA_MAP_ACCESS_WRITE => libc::PROT_WRITE,
+                    RUTABAGA_MAP_ACCESS_RW => libc::PROT_READ | libc::PROT_WRITE,
+                    _ => return Err(ErrUnspec),
+                };
+
+                if offset + resource.size > shm_region.size as u64 {
+                    error!("mapping DOES NOT FIT");
+                }
+                let addr = shm_region.host_addr + offset;
+                debug!(
+                    "mapping: host_addr={:x}, addr={:x}, size={}",
+                    shm_region.host_addr, addr, resource.size
+                );
```

Comment:
Is there a reason why `VirtioShmRegion::host_addr` is not a pointer but a raw integer?

Reply:
i think it's safe to keep it as an integer until the final operation `addr as *mut libc::c_void,` that way the code avoids creating potentially unsafe intermediate pointers.

Reply:
Pointer arithmetic is safe, it's dereferencing that's unsafe.